File: src/auv_description/scripts/cmd_wasd_teleop.py
#!/usr/bin/env python3
import sys, termios, tty, select
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist, Vector3

HELP = """\
AUV keyboard teleop (force/torque only, continuous set & hold)

Keys:
  W/S: ±X   A/D: ±Y   E/Q: ±Z
  I/K: ±Pitch (±Y)   J/L: ±Yaw (±Z)   U/O: ±Roll (±X)
  Space: zero all
"""

def _getch_nonblocking():
    fd = sys.stdin.fileno()
    old = termios.tcgetattr(fd)
    try:
        tty.setraw(fd)
        r, _, _ = select.select([sys.stdin], [], [], 0.0)
        return sys.stdin.read(1) if r else None
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old)

class KeyboardAuvTeleop(Node):
    def __init__(self):
        super().__init__("keyboard_auv_teleop")

        # Params FIRST
        self.declare_parameter("force_topic", "/auve1/force_body")
        self.declare_parameter("torque_topic", "/auve1/torque_body")
        self.declare_parameter("linear_speed", 1.0)
        self.declare_parameter("angular_speed", 1.0)
        self.declare_parameter("rate_hz", 5000.0)
        self.declare_parameter("k_linear", 100.0)
        self.declare_parameter("k_angular", 50.0)

        self.force_topic   = self.get_parameter("force_topic").get_parameter_value().string_value
        self.torque_topic  = self.get_parameter("torque_topic").get_parameter_value().string_value
        self.linear_speed  = float(self.get_parameter("linear_speed").value)
        self.angular_speed = float(self.get_parameter("angular_speed").value)
        self.rate_hz       = max(1.0, float(self.get_parameter("rate_hz").value))
        self.k_linear      = float(self.get_parameter("k_linear").value)
        self.k_angular     = float(self.get_parameter("k_angular").value)

        # Publishers (force/torque only)
        self.pub_force = self.create_publisher(Vector3, self.force_topic, 10)
        self.pub_tau   = self.create_publisher(Vector3, self.torque_topic, 10)
        
        # NOW wait for subscribers (after publishers exist!)
        self.get_logger().info("Waiting for subscribers...")
        while self.pub_force.get_subscription_count() == 0:
            rclpy.spin_once(self, timeout_sec=0.1)
        self.get_logger().info(f"Connected! {self.pub_force.get_subscription_count()} subscriber(s)")
        
        self.get_logger().info(f'Publishing Vector3 force on "{self.pub_force.topic_name}"')
        self.get_logger().info(f'Publishing Vector3 torque on "{self.pub_tau.topic_name}"')
        self.get_logger().info(HELP)

        # Rest of initialization...
        self.twist = Twist()
        self.timer_pub = self.create_timer(1.0 / self.rate_hz, self._on_publish_timer)
        self.stdin_is_tty = sys.stdin.isatty()
        self.timer_key = self.create_timer(0.001, self._on_key_poll)
        self._pub_count = 0

    def _on_publish_timer(self):
        # Map Twist → Vector3 wrench and publish
        f = Vector3(
            x=self.k_linear * self.twist.linear.x,
            y=self.k_linear * self.twist.linear.y,
            z=self.k_linear * self.twist.linear.z,
        )
        tau = Vector3(
            x=self.k_angular * self.twist.angular.x,
            y=self.k_angular * self.twist.angular.y,
            z=self.k_angular * self.twist.angular.z,
        )
        self.pub_force.publish(f)
        self.pub_tau.publish(tau)

        self._pub_count += 1
        if self._pub_count % int(self.rate_hz * 1.0) == 0:
            self.get_logger().info("Publishing force/torque...")

    def _on_key_poll(self):
        if not self.stdin_is_tty:
            return
        ch = _getch_nonblocking()
        if not ch:
            return
        if ch == "\x03":  # Ctrl-C
            rclpy.shutdown()
            return
        self._handle_key(ch)

    def _zero(self):
        self.twist.linear.x = self.twist.linear.y = self.twist.linear.z = 0.0
        self.twist.angular.x = self.twist.angular.y = self.twist.angular.z = 0.0

    def _handle_key(self, ch: str):
        ch = ch.lower()
        updated = True

        # Translation
        if ch == "w":
            self.twist.linear.x = self.linear_speed;  self.twist.linear.y = 0.0
        elif ch == "s":
            self.twist.linear.x = -self.linear_speed; self.twist.linear.y = 0.0
        elif ch == "a":
            self.twist.linear.y = self.linear_speed;  self.twist.linear.x = 0.0
        elif ch == "d":
            self.twist.linear.y = -self.linear_speed; self.twist.linear.x = 0.0
        elif ch == "e":
            self.twist.linear.z = self.linear_speed
        elif ch == "q":
            self.twist.linear.z = -self.linear_speed

        # Rotation (roll=x, pitch=y, yaw=z)
        elif ch == "i":
            self.twist.angular.y = self.angular_speed
        elif ch == "k":
            self.twist.angular.y = -self.angular_speed
        elif ch == "j":
            self.twist.angular.z = self.angular_speed
        elif ch == "l":
            self.twist.angular.z = -self.angular_speed
        elif ch == "u":
            self.twist.angular.x = self.angular_speed
        elif ch == "o":
            self.twist.angular.x = -self.angular_speed

        elif ch == " ":
            self._zero()
        else:
            updated = False

        if updated:
            self.get_logger().info(
                f"Key '{ch}' → lin=({self.twist.linear.x:.2f},"
                f"{self.twist.linear.y:.2f},{self.twist.linear.z:.2f}) "
                f"ang=({self.twist.angular.x:.2f},{self.twist.angular.y:.2f},"
                f"{self.twist.angular.z:.2f})"
            )

def main():
    rclpy.init()
    node = KeyboardAuvTeleop()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.get_logger().info("Shutting down keyboard_auv_teleop.")
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/gncea_description/scripts/cmd_wasd_teleop.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/vector3.hpp>

#include <atomic>
#include <thread>
#include <mutex>
#include <string>
#include <cstdio>
#include <cstdlib>
#include <cerrno>

#include <termios.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/select.h>

using geometry_msgs::msg::Vector3;

struct TwistLite
{
  double lx{0}, ly{0}, lz{0};
  double ax{0}, ay{0}, az{0};
};

class KeyboardAuvTeleop : public rclcpp::Node
{
public:
  KeyboardAuvTeleop()
  : Node("keyboard_gncea_teleop")
  {
    // -------- Parameters ----------
    force_topic_   = declare_parameter<std::string>("force_topic",  "/auve1/force_body");
    torque_topic_  = declare_parameter<std::string>("torque_topic", "/auve1/torque_body");
    linear_speed_  = declare_parameter<double>("linear_speed",  1.0);   // m/s (equivalent)
    angular_speed_ = declare_parameter<double>("angular_speed", 1.0);   // rad/s (equivalent)
    rate_hz_       = std::max(1.0, declare_parameter<double>("rate_hz", 200.0));       // publish rate
    key_poll_hz_   = std::max(50.0, declare_parameter<double>("key_poll_hz", 400.0));  // keyboard poll rate
    k_linear_      = declare_parameter<double>("k_linear",  200.0); // N per (m/s)
    k_angular_     = declare_parameter<double>("k_angular",  100.0); // N·m per (rad/s)

    // -------- QoS (low latency) ----------
    rclcpp::QoS qos(rclcpp::KeepLast(1));
    qos.reliable();
    qos.durability_volatile();

    pub_force_ = create_publisher<Vector3>(force_topic_, qos);
    pub_torque_ = create_publisher<Vector3>(torque_topic_, qos);

    RCLCPP_INFO(get_logger(), "force → \"%s\", torque → \"%s\"", force_topic_.c_str(), torque_topic_.c_str());
    RCLCPP_INFO(get_logger(), "Keys: W/S ±X, A/D ±Y, E/Q ±Z, I/K ±Pitch, J/L ±Yaw, U/O ±Roll, Space zero");

    // -------- Timers ----------
    publish_timer_ = create_wall_timer(
      std::chrono::duration<double>(1.0 / rate_hz_),
      std::bind(&KeyboardAuvTeleop::OnPublishTimer, this));

    // -------- Keyboard thread (/dev/tty fallback) ----------
    start_keyboard_thread();
  }

  ~KeyboardAuvTeleop() override
  {
    stop_keyboard_thread();
  }

private:
  // ---------- Keyboard ----------
  void start_keyboard_thread()
  {
    // Prefer stdin if TTY; else open /dev/tty
    fd_ = isatty(STDIN_FILENO) ? STDIN_FILENO : open("/dev/tty", O_RDONLY | O_NONBLOCK);
    if (fd_ < 0) {
      RCLCPP_ERROR(get_logger(), "Failed to open TTY for keyboard input: %s", strerror(errno));
      return;
    }

    // Save and set raw mode
    if (tcgetattr(fd_, &orig_term_) == 0) {
      termios raw = orig_term_;
      cfmakeraw(&raw);
      tcsetattr(fd_, TCSANOW, &raw);
      raw_ok_ = true;
    } else {
      RCLCPP_WARN(get_logger(), "tcgetattr failed; continuing without raw mode.");
    }

    running_.store(true);
    key_thread_ = std::thread([this]{
      const double period = 1.0 / key_poll_hz_;
      timeval tv{};
      while (running_.load()) {
        fd_set rfds; FD_ZERO(&rfds); FD_SET(fd_, &rfds);
        tv.tv_sec = 0;
        tv.tv_usec = static_cast<suseconds_t>(period * 1e6);
        int ret = select(fd_ + 1, &rfds, nullptr, nullptr, &tv);
        if (ret > 0 && FD_ISSET(fd_, &rfds)) {
          char c;
          ssize_t n = read(fd_, &c, 1);
          if (n == 1) {
            if (c == 0x03) { // Ctrl-C
              rclcpp::shutdown();
              break;
            }
            handle_key(c);
          }
        }
        // if ret==0 (timeout) or ret<0, just loop again
      }
    });

    RCLCPP_INFO(get_logger(), "Reading keys from: %s", (fd_ == STDIN_FILENO) ? "stdin" : "/dev/tty");
  }

  void stop_keyboard_thread()
  {
    running_.store(false);
    if (key_thread_.joinable()) key_thread_.join();

    if (raw_ok_) {
      tcsetattr(fd_, TCSANOW, &orig_term_);
    }
    if (fd_ != STDIN_FILENO && fd_ >= 0) close(fd_);
    fd_ = -1;
  }

  void handle_key(char ch)
  {
    ch = static_cast<char>(::tolower(static_cast<unsigned char>(ch)));
    bool updated = true;
    std::lock_guard<std::mutex> lk(cmd_mtx_);

    // Translation
    if      (ch == 'w') { cmd_.lx =  linear_speed_; cmd_.ly = 0.0; }
    else if (ch == 's') { cmd_.lx = -linear_speed_; cmd_.ly = 0.0; }
    else if (ch == 'a') { cmd_.ly =  linear_speed_; cmd_.lx = 0.0; }
    else if (ch == 'd') { cmd_.ly = -linear_speed_; cmd_.lx = 0.0; }
    else if (ch == 'e') { cmd_.lz =  linear_speed_; }
    else if (ch == 'q') { cmd_.lz = -linear_speed_; }
    // Rotation (roll=x, pitch=y, yaw=z)
    else if (ch == 'u') { cmd_.ax =  angular_speed_; }
    else if (ch == 'o') { cmd_.ax = -angular_speed_; }
    else if (ch == 'i') { cmd_.ay =  angular_speed_; }
    else if (ch == 'k') { cmd_.ay = -angular_speed_; }
    else if (ch == 'j') { cmd_.az =  angular_speed_; }
    else if (ch == 'l') { cmd_.az = -angular_speed_; }
    else if (ch == ' ') { zero_cmd_locked(); }
    else { updated = false; }

    if (updated) {
      RCLCPP_INFO_THROTTLE(
        get_logger(), *get_clock(), 500 /*ms*/,
        "lin=(%.2f,%.2f,%.2f) ang=(%.2f,%.2f,%.2f)",
        cmd_.lx, cmd_.ly, cmd_.lz, cmd_.ax, cmd_.ay, cmd_.az);
    }
  }

  void zero_cmd_locked()
  {
    cmd_.lx = cmd_.ly = cmd_.lz = 0.0;
    cmd_.ax = cmd_.ay = cmd_.az = 0.0;
  }

  // ---------- Publish ----------
  void OnPublishTimer()
  {
    TwistLite copy;
    {
      std::lock_guard<std::mutex> lk(cmd_mtx_);
      copy = cmd_;
    }

    Vector3 f, tau;
    f.x   = k_linear_  * copy.lx;
    f.y   = k_linear_  * copy.ly;
    f.z   = k_linear_  * copy.lz;
    tau.x = k_angular_ * copy.ax;
    tau.y = k_angular_ * copy.ay;
    tau.z = k_angular_ * copy.az;

    pub_force_->publish(f);
    pub_torque_->publish(tau);
  }

private:
  // Parameters
  std::string force_topic_, torque_topic_;
  double linear_speed_{1.0}, angular_speed_{1.0};
  double rate_hz_{200.0}, key_poll_hz_{400.0};
  double k_linear_{100.0}, k_angular_{50.0};

  // Publishers
  rclcpp::Publisher<Vector3>::SharedPtr pub_force_;
  rclcpp::Publisher<Vector3>::SharedPtr pub_torque_;

  // Timers
  rclcpp::TimerBase::SharedPtr publish_timer_;

  // Command state
  TwistLite cmd_;
  std::mutex cmd_mtx_;

  // Keyboard thread
  std::thread key_thread_;
  std::atomic<bool> running_{false};
  int fd_{-1};
  termios orig_term_{};
  bool raw_ok_{false};
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<KeyboardAuvTeleop>();
  rclcpp::executors::SingleThreadedExecutor exec;
  exec.add_node(node);
  try {
    exec.spin();
  } catch (...) {
    // fallthrough to shutdown
  }
  rclcpp::shutdown();
  return 0;
}
